src/Package/Impl/Repl/Commands/SourceFileHelper.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Microsoft.R.Components.InteractiveWorkflow;
+using Microsoft.R.Host.Client;
+using Microsoft.VisualStudio.R.Package.Shell;
+using Microsoft.VisualStudio.R.Package.Utilities;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.R.Package.Repl.Commands {
+    internal static class SourceFileHelper {
+        public static void SourceFiles(IEnumerable<string> files) {
+            VsAppShell.Current.AssertIsOnMainThread();
+
+            bool debugging = IsDebugging();
+            var operations = GetOperations();
+            var list = new List<string>();
+
+            Task.Run(async () => {
+                foreach (var f in files) {
+                    list.Add(await SessionUtilities.GetRShortenedPathNameAsync(f));
+                }
+                VsAppShell.Current.DispatchOnUIThread(() => {
+                    foreach (var filePath in list) {
+                        operations.EnqueueExpression($"{(debugging ? "rtvs::debug_source" : "source")}({filePath.ToRStringLiteral()})", addNewLine: true);
+                     }
+                });
+            });
+        }
+
+        public static void SourceFile(string file) {
+            VsAppShell.Current.AssertIsOnMainThread();
+
+            bool debugging = IsDebugging();
+            var operations = GetOperations();
+
+            Task.Run(async () => {
+                file = await SessionUtilities.GetRShortenedPathNameAsync(file);
+                VsAppShell.Current.DispatchOnUIThread(() => {
+                    operations.ExecuteExpression($"{(debugging ? "rtvs::debug_source" : "source")}({file.ToRStringLiteral()})");
+                });
+            });
+        }
+
+        private static bool IsDebugging() {
+            var debugger = VsAppShell.Current.GetGlobalService<IVsDebugger>(typeof(SVsShellDebugger));
```

Comment:
Although I guess the real question that this code should ask is, "are breakpoints enabled", in which case `DebugSession` can track that (it doesn't currently - it just passes it on to R code which does the actual tracking - but it would be easy to add).